### PR TITLE
Adds a basic unit test for PointCloud functionality

### PR DESCRIPTION
**Pull request by arjo129**
This PR adds a very trivial (happy path) unit test for PointCloud functionality. Hopefully, I will get more time to add more functionality to this test over time.

TODOS:
* Check behaviour with NaNs
* Check mismatched float and pointcloud sizes
* Check malformed point cloud.

Signed-off-by: Arjo Chakravarty <arjo@openrobotics.org>

### CODE REVIEW

File: test/integration/point_cloud.cc [hunk 1]
```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright (C) 2021 Open Source Robotics Foundation
```
Comment (ahcorde) [inline test/integration/point_cloud.cc:2]:
```suggestion
 * Copyright (C) 2022 Open Source Robotics Foundation
```

File: test/integration/point_cloud.cc [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+*/
+
+#include <gtest/gtest.h>
+#include <string>
+
+#include <gz/msgs/world_stats.pb.h>
+#include <gz/msgs/marker.pb.h>
+#include <gz/msgs/material.pb.h>
+#include <gz/msgs/float_v.pb.h>
+#include <gz/msgs/pointcloud_packed.pb.h>
```
Comment (ahcorde) [inline test/integration/point_cloud.cc:25]: alphabetize

File: test/integration/point_cloud.cc [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+};
+
+using namespace std::chrono_literals;
+
+using namespace gz;
+using namespace gui;
+
+class PointCloudTestFixture : public ::testing::Test
+{
+
```
Comment (ahcorde) [inline test/integration/point_cloud.cc:56]:
```suggestion
```

File: test/integration/point_cloud.cc [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+      this->InitMockData();
+
+      this->node.Advertise("/marker",
+        &PointCloudTestFixture::OnMarkerMsg, this);
+      this->receivedMsg = false;
+    }
+
+    public: void InitMockData()
+    {
+
```
Comment (ahcorde) [inline test/integration/point_cloud.cc:82]:
```suggestion
```

File: test/integration/point_cloud.cc [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+      }
+      else if (_req.action() == gz::msgs::Marker::ADD_MODIFY)
+      {
+        EXPECT_EQ(_req.id(), 1);
+        EXPECT_EQ(_req.ns(), "/point_cloud/flat");
+        EXPECT_EQ(_req.type(), gz::msgs::Marker::POINTS);
+        EXPECT_EQ(_req.visibility(), gz::msgs::Marker::GUI);
+        if (_req.point().size() != 0)
+        {
+          // We might recieve empty packets as the sending process
```
Comment (ahcorde) [inline test/integration/point_cloud.cc:142]:
```suggestion
          // We might receive empty packets as the sending process
```

File: test/integration/point_cloud.cc [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+
+/////////////////////////////////////////////////
+TEST_F(PointCloudTestFixture,
+  GZ_UTILS_TEST_ENABLED_ONLY_ON_LINUX(PointCloudTestFixture))
+{
+  common::Console::SetVerbosity(4);
+
+  // Load the plugin
+  Application app(g_argc, g_argv);
+  app.AddPluginPath(std::string(PROJECT_BINARY_PATH) + "/lib");
```
Comment (ahcorde) [inline test/integration/point_cloud.cc:164]:
```suggestion
  app.AddPluginPath(common::joinPaths(std::string(PROJECT_BINARY_PATH), "lib"));
```

File: test/integration/point_cloud.cc [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+  auto window = app.findChild<MainWindow *>();
+  ASSERT_NE(window, nullptr);
+
+  // Get plugin
+  auto plugins = window->findChildren<Plugin *>();
+  EXPECT_EQ(plugins.size(), 1);
+
+  // Show, but don't exec, so we don't block
+  window->QuickWindow()->show();
+
```
Comment (ahcorde) [inline test/integration/point_cloud.cc:188]:
```suggestion
```

**Review by ahcorde — CHANGES_REQUESTED** [review]

**Comment by jennuine** [comment]
@arjo129 what is the status of this PR? Will you be addressing the TODOs here?

**Comment by arjo129** [comment]
We can merge this as is and Ill increase coverage over time.

File: test/integration/point_cloud.cc [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+      this->InitMockData();
+
+      this->node.Advertise("/marker",
+        &PointCloudTestFixture::OnMarkerMsg, this);
+      this->receivedMsg = false;
+    }
+
+    public: void InitMockData()
+    {
+      gz::msgs::InitPointCloudPacked(pcMsg, "some_frame", true,
```
Comment (jennuine) [inline test/integration/point_cloud.cc:81]: nit: adding `this->` to member variables (below as well)

```suggestion
      gz::msgs::InitPointCloudPacked(this->pcMsg, "some_frame", true,
```

File: test/integration/point_cloud.cc [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+      this->flatPub.Publish(flatMsg);
+    }
+
+    /// \brief Callback that receives marker messages.
+    /// \param[in] _req The marker message.
+    public: void OnMarkerMsg(const gz::msgs::Marker &_req)
+    {
+      if (_req.action() == gz::msgs::Marker::DELETE_ALL)
+      {
+        // TODO(arjo129): Don't clear all markers only one marker.
```
Comment (jennuine) [inline test/integration/point_cloud.cc:130]: Is this a feature `TODO` (ie. in the core code, not in this test)? If so, can you create an issue for better visibility?

File: test/integration/point_cloud.cc [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+  EXPECT_EQ(plugins.size(), 1);
+
+  // Show, but don't exec, so we don't block
+  window->QuickWindow()->show();
+
+  this->Publish();
+
+  int sleep = 0;
+  int maxSleep = 30;
+  while (sleep < maxSleep)
```
Comment (jennuine) [inline test/integration/point_cloud.cc:192]:
```suggestion
  while (sleep < maxSleep && !this->receivedMsg)
```

**Review by jennuine — COMMENTED** [review]
Could this test be added to [PointCloud_TEST](https://github.com/gazebosim/gz-gui/blob/gz-gui7/src/plugins/point_cloud/PointCloud_TEST.cc) instead?

File: test/integration/point_cloud.cc (hunk 9 above)
Comment (arjo129) [reply test/integration/point_cloud.cc:130]: Ah this comment doesn't belong here.

File: src/plugins/point_cloud/PointCloud_TEST.cc [hunk 11]
```diff
@@ -18,35 +18,156 @@
...
+        }
+      }
+      else if(_req.action() == gz::msgs::Marker::DELETE_ALL)
+      {
+        // Do nothing. Its ok to clear the screen.
+      }
+      else
+      {
+        ADD_FAILURE();
+      }
```
Comment (jennuine) [inline src/plugins/point_cloud/PointCloud_TEST.cc:145]:
```suggestion
      else if (_req.action() != msgs::Marker::DELETE_ALL)
      {
        // if DELETE_ALL, its ok to clear the screen. Otherwise fail
        ADD_FAILURE();
      }
```

File: src/plugins/point_cloud/PointCloud_TEST.cc [hunk 12]
```diff
@@ -18,35 +18,156 @@
...
 #include <gz/common/Filesystem.hh>
+#include <gz/rendering/RenderEngine.hh>
+#include <gz/rendering/RenderingIface.hh>
+#include <gz/rendering/Scene.hh>
+#include <gz/transport/Node.hh>
 #include <gz/utils/ExtraTestMacros.hh>
 
 #include "test_config.hh"  // NOLINT(build/include)
 #include "gz/gui/Application.hh"
+#include "gz/gui/GuiEvents.hh"
```
Comment (jennuine) [inline src/plugins/point_cloud/PointCloud_TEST.cc:35]: These headers aren't needed
```suggestion
#include <gz/transport/Node.hh>
#include <gz/utils/ExtraTestMacros.hh>

#include "test_config.hh"  // NOLINT(build/include)
#include "gz/gui/Application.hh"
```

File: src/plugins/point_cloud/PointCloud_TEST.cc [hunk 13]
```diff
@@ -18,35 +18,156 @@
...
 #include "gz/gui/MainWindow.hh"
 #include "gz/gui/Plugin.hh"
-#include "PointCloud.hh"
```
Comment (jennuine) [inline src/plugins/point_cloud/PointCloud_TEST.cc:29]: Preferably, I would leave the `PointCloud.hh` header and remove the `gz/gui/Plugin.hh` one

File: src/plugins/point_cloud/PointCloud_TEST.cc [hunk 14]
```diff
@@ -18,35 +18,156 @@
...
+        EXPECT_EQ(_req.ns(), "/point_cloud/flat");
+        EXPECT_EQ(_req.type(), msgs::Marker::POINTS);
+        EXPECT_EQ(_req.visibility(), msgs::Marker::GUI);
+        if (_req.point().size() != 0)
+        {
+          // We might receive empty packets as the sending process
+          // is asynchronuous
+          EXPECT_EQ(_req.point().size(), this->flatMsg.data().size());
+          EXPECT_EQ(_req.materials().size(), this->flatMsg.data().size());
+          this->receivedMsg = true;
```
Comment (jennuine) [inline src/plugins/point_cloud/PointCloud_TEST.cc:135]: It would be good to check the data matches as well

**Review by jennuine — COMMENTED** [review]
Left some more minor comments. Could you summarize the reason for the point cloud publisher/msg? In the marker message callback, only the `flatMsg` data is being checked but not `pcMsg`. It would be nice to check the point cloud data as well.

File: src/plugins/point_cloud/PointCloud_TEST.cc (hunk 14 above)
Comment (arjo129) [reply src/plugins/point_cloud/PointCloud_TEST.cc:174]: [c1f232b](https://github.com/gazebosim/gz-gui/pull/496/commits/c1f232bf5eaf8ab3b7865eeb809cb9899905b614)

**Comment by arjo129** [comment]
> Could you summarize the reason for the point cloud publisher/msg? In the marker message callback, only the flatMsg data is being checked but not pcMsg. It would be nice to check the point cloud data as well.

The rationale for this is that the pcMsg provides data about position of the markers whereas the marker messages provide data about position. The test checks that the marker messages have the correct color values (which is what the point-cloud plugin basically does).

File: src/plugins/point_cloud/PointCloud_TEST.cc [hunk 15]
```diff
@@ -37,16 +41,170 @@ char* g_argv[] =
...
+          // is asynchronuous
+          EXPECT_EQ(_req.point().size(), this->flatMsg.data().size());
+          EXPECT_EQ(_req.materials().size(), this->flatMsg.data().size());
+
+          auto dR = maxColor.R() - minColor.R();
+          auto dG = maxColor.G() - minColor.G();
+          auto dB = maxColor.B() - minColor.B();
+          auto dA = maxColor.A() - minColor.A();
+
+          for (std::size_t idx = 0; idx < _req.point().size(); idx++)
```
Comment (iche033) [inline src/plugins/point_cloud/PointCloud_TEST.cc:143]: compile warning here:

```
PointCloud_TEST.cc:143:41: warning: comparison of integer expressions of different signedness: ‘std::size_t’ {aka ‘long unsigned int’} and ‘int’ [-Wsign-compare]
  143 |           for (std::size_t idx = 0; idx < _req.point().size(); idx++)
```

Comment (iche033) [reply src/plugins/point_cloud/PointCloud_TEST.cc]: [9c5e3b6](https://github.com/gazebosim/gz-gui/pull/496/commits/9c5e3b6c7d23cfb55663017b611b9b24ba35b507)

**Review by iche033 — APPROVED** [review]

**Comment by mjcarroll** [comment]
I think that the `focal` failures here in CI are unrelated to this PR?

**Review by jennuine — APPROVED** [review]
